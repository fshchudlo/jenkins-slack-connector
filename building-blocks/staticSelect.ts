import { textBlock } from "./textBlock";
import { placeholder } from "./placeholder";

export function staticSelect(label: string, options: Record<string, string>, action_id: string, initial_option: string | null = null, placeholder_text: string | null = null) {
    if (Object.keys(options).length > 100) {
        throw new Error("Maximum number of options allowed by Slack is 100");
    }

    const result = {
        type: "input",
        label: textBlock(label),
        block_id: action_id,
        element: {
            type: "static_select",
            options: Object.entries(options).map(([key, value]) => selectOption(key, value)),
            action_id: action_id
        }
    };

    if (initial_option !== null) {
        (<any>result.element).initial_option = selectOption(initial_option, options[initial_option]);
    }

    if (placeholder_text !== null) {
        (<any>result.element).placeholder = placeholder(placeholder_text);
    }

    return result;
}

function selectOption(value: string, text: string | null = null) {
    const textToDisplay = text === null ? value : text;
    if (textToDisplay.length > 75) {
        throw new Error("Display text length allowed by Slack is 75");
    }
    return {
        text: textBlock(textToDisplay),
        value: value
    };
}