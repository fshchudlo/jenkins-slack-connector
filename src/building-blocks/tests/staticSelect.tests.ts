import { staticSelect } from "../staticSelect";

describe("staticSelect", () => {
    it("should create a static select with the correct properties", () => {
        const options = { option1: "Option 1", option2: "Option 2" };
        const control = staticSelect("Select an option", options, "action_id", "option1", "Choose an option");
        expect(control).toMatchSnapshot();
    });

    it("should throw an error if more than 100 options are provided", () => {
        const options: { [key: string]: string } = {};
        for (let i = 1; i <= 101; i++) {
            options[`option${i}`] = `Option ${i}`;
        }
        expect(() => staticSelect("Select an option", options, "some_action_id")).toThrow(
            "Maximum number of options allowed by Slack is 100"
        );
    });

    it("should throw an error if the display text length exceeds 75 characters", () => {
        const options = { option1: "This is a very long option text that exceeds the maximum allowed length of 75 symbols" };
        expect(() => staticSelect("Select an option", options, "some_action_id")).toThrow(
            "Display text length allowed by Slack is 75"
        );
    });
});
