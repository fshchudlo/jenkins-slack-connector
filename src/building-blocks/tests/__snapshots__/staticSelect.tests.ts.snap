// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`staticSelect should create a static select with the correct properties 1`] = `
{
  "block_id": "action_id",
  "element": {
    "action_id": "action_id",
    "initial_option": {
      "text": {
        "text": "Option 1",
        "type": "plain_text",
      },
      "value": "option1",
    },
    "options": [
      {
        "text": {
          "text": "Option 1",
          "type": "plain_text",
        },
        "value": "option1",
      },
      {
        "text": {
          "text": "Option 2",
          "type": "plain_text",
        },
        "value": "option2",
      },
    ],
    "placeholder": {
      "text": "Choose an option",
      "type": "plain_text",
    },
    "type": "static_select",
  },
  "label": {
    "text": "Select an option",
    "type": "plain_text",
  },
  "type": "input",
}
`;
